cmake: Improve safety and robustness during building `crc32c` subtree

hebasto opened the pull request:
It was [agreed](https://github.com/bitcoin/bitcoin/pull/30905#issuecomment-2426263301) not to adopt the upstream buildsystem the `crc32c` subtree.

So this PR ensures that our custom build script properly handles potentially harmful situations, including:
1. The upstream buildsystem changes remaining unnoticed during the subtree update, even when some of those changes need to be reflected in `crc32c.cmake`.

2. The order of `include(cmake/introspection.cmake)` and `include(cmake/crc32c.cmake)` being inadvertently altered during refactoring or other changes in the main buildsystem.

hebasto committed:
cmake, refactor: Add `crc32c_subtree_dir` helper variable

This change minimizes repetitions and improves readability.

hebasto committed:
cmake, crc32c: Improve check robustness

This change prevents accidental breakage of the compiled code when
moving code around `include(cmake/crc32c.cmake)`.

purpleKarrot commented:
Whether the upstream `CMakeLists.txt` file is used or not, the `crc32c` subtree should be added with `add_subdirectory` and not `include`. The latter prevents variables from leaking into the parent directory.

fanquake commented on cmake/crc32c.cmake:
```diff
@@ -2,11 +2,26 @@
+# See the discussion in https://github.com/bitcoin/bitcoin/pull/30905.
+# However, we want to ensure that no changes upstream remain unnoticed.
+file(SHA256 ${crc32c_subtree_dir}/CMakeLists.txt upstream_cmakelists_hash)
+if(NOT upstream_cmakelists_hash STREQUAL "f5c90773b3afd4ee95e792356a802758a29a3751878d0388c555c575c19ed4ce")
```
Why does this file sha256sum to something different in the Windows builds: https://github.com/bitcoin/bitcoin/actions/runs/13099284830/job/36545477809?pr=31779#step:8:925:
```bash
 CMake Error at cmake/crc32c.cmake:12 (message):
  D:/a/bitcoin/bitcoin/src/crc32c/CMakeLists.txt has been modified.
Call Stack (most recent call first):
  CMakeLists.txt:594 (include)

-- Configuring incomplete, errors occurred!
```

hebasto replied:
I think this happens due to the [Windows-style line endings](https://stackoverflow.com/questions/10418975/how-to-change-line-ending-settings).

hebasto committed:
cmake, crc32c: Check for upstream buildsystem changes

This change ensures that no upstream modifications go unnoticed during
the crc32c subtree update. Certain changes, such as conditionally
including a new source file, must be reflected in the `crc32c.cmake`
script.

hebasto commented:
@purpleKarrot 
> Whether the upstream `CMakeLists.txt` file is used or not, the `crc32c` subtree should be added with `add_subdirectory` and not `include`. The latter prevents variables from leaking into the parent directory.

Certainly, I agree with keeping things as scoped as possible.

The initial plan was to replace `include(cmake/crc32c.cmake)` with `add_subdirectory(src/crc32c)`. However, the perspective has since [changed](https://github.com/bitcoin/bitcoin/pull/30905#issuecomment-2426263301).

Your suggestion is not specific to crc32c alone but applies to other subtrees as well. It seems beyond the scope of this PR and might be better suited for a separate one, no?

fanquake commented:
~0.

> However, the perspective has since https://github.com/bitcoin/bitcoin/pull/30905#issuecomment-2426263301.

#30905 wasn't merged because it just took upstreams compilation flags, which didn't make sense for us, and because switching came with pretty much no benefit (and still couldn't be done without other modifications to the subtree).

I think time would be better spent modifying our subtree, so we could use the CMake file (if that's what we really want to do), rather than trying to implement changes here to detect situations that are unlikely to occur (especially given the state of upstream development) and would just be prevented by switching.

fanquake commented:
Another thought, isn't this going to break git bisect, or, result in having to do unclean subtree pulls? As the commit where you update the subtree will no-longer compile unless it also simultaneously changes the hash hardcoded into our build system.

hebasto commented:
> Another thought, isn't this going to break git bisect, or, result in having to do unclean subtree pulls? As the commit where you update the subtree will no-longer compile unless it also simultaneously changes the hash hardcoded into our build system.

Fair enough.
